### PR TITLE
RemoteFX software decoding

### PR DESCRIPTION
**Pull request by llyzs**
This pull request adds a new library for RemoteFX software decoding called librfx, and use it in xfreerdp.

### CODE REVIEW

File: X11/xf_decode.c [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+	{
+		case XF_CODEC_REMOTEFX:
+
+			message = rfx_process_message((RFX_CONTEXT *) xfi->rfx_context, bitmapData, bitmapDataLength);
+
+			XSetFunction(xfi->display, xfi->gc, GXcopy);
+			XSetFillStyle(xfi->display, xfi->gc, FillSolid);
+			XSetClipRectangles(xfi->display, xfi->gc, x, y, (XRectangle*)message->rects, message->num_rects, YXBanded);
+
+			for (i = 0; i < message->num_tiles; i++)
```
Comment (awakecoding) [inline X11/xf_decode.c:69]: I'd be interested in adapting this code for usage in DirectFB. Can you add just a few comments on what's the difference between message->tiles[] and message->rects[] exactly?

File: X11/xf_decode.c [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+void
+xf_decode_data(xfInfo * xfi, uint8 * data, int data_size)
+{
+	uint16 cmdType;
+	uint32 bitmapDataLength;
+	int destLeft;
+	int destTop;
+	int size;
+
+	///printf("xf_decode_data: %d\n", data_size);
```
Comment (awakecoding) [inline X11/xf_decode.c]: Maybe cleanup commented out printf's a little bit, or add a new DEBUG_RFX

File: X11/xf_types.h [hunk 3]
```diff
@@ -33,6 +33,12 @@
 #define SET_XFI(_inst, _xfi) (_inst)->param1 = _xfi
 #define GET_XFI(_inst) ((xfInfo *) ((_inst)->param1))
 
+enum
+{
+	XF_CODEC_NONE,
+	XF_CODEC_REMOTEFX
```
Comment (awakecoding) [inline X11/xf_types.h:39]: Besides having RemoteFX or no RemoteFX, is there any other "codec" option? Should be abstract GDI vs RemoteFX as being "no codec" vs "codec is RemoteFX", or could we simply say it's an on/off switch for RemoteFX?

File: librfx/rfx_test.c [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+*/
+
+/*
+   This file is only a simple unit testing program and is NOT part of librfx.
+   The sample data comes from [MS-RDPRFX] 4.2.3, which is decoded into three
+   vertical bands in red (21x64), green (23x64) and blue(20x64) color.
+
+   Compiles with:
+
+   gcc -Wall -o rfx_test rfx_test.c -lrfx
```
Comment (awakecoding) [inline librfx/rfx_test.c]: This appears to be some custom unit test, which is put together with the actual library. Could it be modified to use cunit and moved to the cunit folder instead?

**Comment by awakecoding** [comment]
I tried it today on my macbook air, and it worked out pretty well. I'd like to extend libgdi a little bit to be able to automatically support RemoteFX decoding instead of the X11-specific xf_decode.c.

File: X11/xf_decode.c (hunk 1 above)
Comment (llyzs) [inline X11/xf_decode.c:69]: OK I have added some comments here and also in the API header rfx.h

File: X11/xf_decode.c (hunk 2 above)
Comment (llyzs) [inline X11/xf_decode.c]: Removed.

File: X11/xf_types.h (hunk 3 above)
Comment (llyzs) [inline X11/xf_types.h:39]: Actually yes, as I replied in the mailing list.

File: librfx/rfx_test.c (hunk 4 above)
Comment (llyzs) [inline librfx/rfx_test.c]: Done.

**Pull request merged**